### PR TITLE
raft: Add applied index as an argument to newRaft and RestartNode.

### PR DESCRIPTION
**Pull request by bdarnell**
This PR replaces #2121 and is based on comments there. 

Closes cockroachdb/etcd#6.

@xiang90 @yichengq

### CODE REVIEW

File: raft/raft.go [hunk 1]
```diff
@@ -172,6 +173,9 @@ func newRaft(id uint64, peers []uint64, election, heartbeat int, storage Storage
 	if !isHardStateEqual(hs, emptyState) {
 		r.loadState(hs)
 	}
+	if applied > 0 {
```
Comment (xiang90) [inline raft/raft.go]: print out applied at line 186

**Comment by xiang90** [comment]
lgtm

File: raft/raft.go (hunk 1 above)
Comment (bdarnell) [inline raft/raft.go]: The applied parameter as passed in, or `raftlog.applied` (which is effectively `max(applied, raftlog.firstIndex())`)?

Comment (xiang90) [inline raft/raft.go]: if the passed in applied is not zero, a correct applied should always be greater than raftlog.firstIndex(). right?

so i prefer raftlog.applied.

Comment (bdarnell) [inline raft/raft.go]: Correct. Raft will panic if you pass an applied index that is less than firstIndex (or greater than HardState.Commit)

File: raft/raft.go [hunk 2]
```diff
@@ -179,8 +183,8 @@ func newRaft(id uint64, peers []uint64, election, heartbeat int, storage Storage
 		nodesStrs = append(nodesStrs, fmt.Sprintf("%x", n))
 	}
 
-	log.Printf("raft: newRaft %x [peers: [%s], term: %d, commit: %d, lastindex: %d, lastterm: %d]",
-		r.id, strings.Join(nodesStrs, ","), r.Term, r.raftLog.committed, r.raftLog.lastIndex(), r.raftLog.lastTerm())
+	log.Printf("raft: newRaft %x [peers: [%s], term: %d, commit: %d, applied: %d, lastindex: %d, lastterm: %d]",
```
Comment (bdarnell) [inline raft/raft.go]: I've updated the log message.

**Pull request merged**